lib/HLSL/HLLowerUDT.cpp

```diff
@@ -302,14 +285,17 @@ void hlsl::ReplaceUsesForLoweredUDT(Value *V, Value *NewV) {
         // Replace bitcast argument with new value
         use.set(NewV);
       }
-
     } else if (ConstantExpr *CE = dyn_cast<ConstantExpr>(user)) {
       // Constant AddrSpaceCast, or BitCast
       if (CE->getOpcode() == Instruction::AddrSpaceCast) {
-        ReplaceUsesForLoweredUDT(
-            user,
-            InsertAddrSpaceCastIfRequired(
-                cast<Constant>(NewV), CE->getType()->getPointerAddressSpace()));
+        DXASSERT(
+            CE->getType()->getPointerAddressSpace() != NewAddrSpace &&
+                OriginalAddrSpace == NewAddrSpace,
+            "When replace Constant, V and NewV must have same address space");
+        Constant *NewAC = ConstantExpr::getAddrSpaceCast(
```

Comment:
I'll admit to being a bit confused by `ConstantExpr::getAddrSpaceCast` but it looks like rather than return the original value when they match, barring some special cases that have nothing to do with type matching, it will return nullptr for `AddrSpaceCast`? 😮 
https://github.com/microsoft/DirectXShaderCompiler/blob/52ace1c57cb782f02ae8ae7fe011bf9c5258a295/lib/IR/ConstantFold.cpp#L723

Reply:
The DXASSERT about Constant *NewAC = ConstantExpr::getAddrSpaceCast should make sure the types don't match.

Reply:
@pow2clk:
> it will return nullptr for AddrSpaceCast? 😮

The `return nullptr` in `ConstantFoldCastInstruction` you pointed out isn't directly returned by `getAddrSpaceCast`, and `getAddrSpaceCast` only returns `nullptr` if you pass `true` in optional `OnlyIfReduced` argument.

https://github.com/microsoft/DirectXShaderCompiler/blob/52ace1c57cb782f02ae8ae7fe011bf9c5258a295/lib/IR/Constants.cpp#L1577-L1581